### PR TITLE
issue/6211-support-not-in

### LINKED ISSUE #6211: Support "not in"
The following code is not supported:

```
if value not in items:

end
```

This has to be written as:

```
if not (value in items):

end
```

### PR DESCRIPTION
**Pull request by FloLey**
# Description

The following code is not supported:

```
if value not in items:

end
```
This has to be written as:

```
if not (value in items):

end
```
with this PR both options are supported and do the same thing.

closes #6211 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: changelogs/unreleased/6211-support-not-in,.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+description: Add the syntactic sugar "value not in items" for "not (value in items)"
+issue-nr: 6202
+change-type: minor
+destination-branches: [master, iso6]
+sections:
+  bugfix: "{{description}}"
```
Comment (sanderr) [inline changelogs/unreleased/6211-support-not-in,.yml:6]: Definitely not a bugfix. I'd label it minor improvement.

File: changelogs/unreleased/6211-support-not-in,.yml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+description: Add the syntactic sugar "value not in items" for "not (value in items)"
```
Comment (sanderr) [inline changelogs/unreleased/6211-support-not-in,.yml:1]: I'm not sure the "syntactic sugar" part is relevant for the end user. I'd just call it the `not in` operator or something.

File: tests/compiler/test_not.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+def test_list_not_in(snippetcompiler, capsys) -> None:
+    """
+    Verify the basic workings of the list comprehension expression.
+    """
+    snippetcompiler.setup_for_snippet(
+        textwrap.dedent(
+            """
+            test_list = [1, 2, 3, 4, 5]
+
+            if 6 not in test_list:
```
Comment (sanderr) [inline tests/compiler/test_not.py:59]: This test isn't very extensive. I would also add a check for a value that is in fact in there. You could use the language's single-value semantics to do assertions if you want to keep it verbose:
```
assert = true
assert = not (6 in test_list)
assert = 6 not in test_list
assert = 5 in test_list
assert = not (5 not in test_list)
...
```

**Review by sanderr — APPROVED** [review]
When I run a compile I get `WARNING: 3 shift/reduce conflicts`.

File: changelogs/unreleased/6211-support-not-in,.yml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+description: Add the "not in" operator
+issue-nr: 6202
```
Comment (Hugo-Inmanta) [inline changelogs/unreleased/6211-support-not-in,.yml:2]:
```suggestion
issue-nr: 6211
```

File: changelogs/unreleased/6211-support-not-in,.yml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+description: Add the "not in" operator
```
Comment (Hugo-Inmanta) [inline changelogs/unreleased/6211-support-not-in,.yml:1]:
```suggestion
description: Add the "not in" operator.
```

**Review by Hugo-Inmanta — APPROVED** [review]
Minor changes on the changelog other than that looks good to me :+1:

File: tests/compiler/test_not.py (hunk 3 above)
Comment (Hugo-Inmanta) [reply tests/compiler/test_not.py]: cool stuff !

File: changelogs/unreleased/6211-support-not-in,.yml (hunk 4 above)
Comment (FloLey) [reply changelogs/unreleased/6211-support-not-in,.yml:2]: again... i always mix those up, thanks.

**Review by sanderr — CHANGES_REQUESTED** [review]
Meant to request changes earlier

File: changelogs/unreleased/6211-support-not-in,.yml [hunk 6]
```diff
@@ -1,6 +1,6 @@
-description: Add the syntactic sugar "value not in items" for "not (value in items)"
+description: Add the "not in" operator
```
Comment (sanderr) [inline changelogs/unreleased/6211-support-not-in,.yml]: This is possibly invalid yaml, you may have to quote the entire thing in single quotes. I'm not sure. The merge tool will tell you if it's invalid.

**Review by sanderr — APPROVED** [review]

**Review by sanderr — CHANGES_REQUESTED** [review]
Forgot about the shift-reduce conflicts

**Review by sanderr — APPROVED** [review]

**Pull request closed without merging**